src/main/java/gov/usgs/wma/waterdata/groundwater/S3Bucket.java

```diff
@@ -38,10 +37,11 @@ public String getKeyName() {
 	@Override
 	public void close() throws Exception {
 		try {
-			writer.close();
-			sendS3();
+			if (writer != null) {
+				writer.close();
+			}
```

Comment:
I think there is a bug by letting 'sendS3' be called implicitly by close, so I made it explicit.

The auto-close construct allows a try block to ensure that 'close' is always called on a resource, regardless of if an error happens.  Its the equivalent of calling close in a finally block.
BUT, in this case we don't want 'sendS3' to be called if there is an error.  We might hit an exception when we are only be half way through writing a row and/or half way through generating the file.  If we send a partial file, the code on the receiving end is assuming that each file is a 'full', so it will delete any missing records for that state.

Thus, I removed the 'send side effect' of close and send is now called explicitly where it is needed in the code.

Reply:
The code works in either way. I prefer the S3Bucket managing the temp file. One other reason is that if AWS gets with the program and gives us a real stream then the S3Bucket class is the only one that need know and make the change. Any calling class need not be the wiser. In this impl, it does need to know. Now, you might be thinking. "This is in one small microservice that is self contained." Well, it might be right now but it might be lifted out for another one. If we keep it self contained, again the benefit of the self contained temp file is a bonus.